cli/flox-activations/src/cli/activate.rs

```diff
@@ -0,0 +1,219 @@
+
+pub const FLOX_SERVICES_TO_START_VAR: &str = "_FLOX_SERVICES_TO_START";
+pub const FLOX_ACTIVATE_START_SERVICES_VAR: &str = "FLOX_ACTIVATE_START_SERVICES";
+
+impl ActivateArgs {
+    pub fn handle(self) -> Result<(), anyhow::Error> {
+        let contents = fs::read_to_string(&self.activate_data)?;
+        let data: ActivateCtx = serde_json::from_str(&contents)?;
+
+        fs::remove_file(&self.activate_data)?;
+
+        let activate_script_command = Self::assemble_command_for_activate_script(data.clone());
+        // when output is not a tty, and no command is provided
+        // we just print an activation script to stdout
+        //
+        // That script can then be `eval`ed in the current shell,
+        // e.g. in a .bashrc or .zshrc file:
+        //
+        //    eval "$(flox activate)"
+        if data.invocation_type == InvocationType::InPlace {
+            Self::activate_in_place(activate_script_command, data.shell)?;
+
+            return Ok(());
+        }
+
+        // These functions will only return if exec fails
+        if data.invocation_type == InvocationType::Interactive {
+            Self::activate_interactive(activate_script_command)
+        } else {
+            Self::activate_command(activate_script_command, data.run_args)
+        }
+    }
+
+    /// Used for `flox activate -- run_args`
+    fn activate_command(mut activate_script_command: Command, run_args: Vec<String>) -> Result<()> {
+        // The activation script works like a shell in that it accepts the "-c"
+        // flag which takes exactly one argument to be passed verbatim to the
+        // userShell invocation. Take this opportunity to combine these args
+        // safely, and *exactly* as the user provided them in argv.
+        activate_script_command
+            .arg("-c")
+            .arg(Self::quote_run_args(&run_args));
+
+        debug!("running activation command: {:?}", activate_script_command);
+
+        // exec should never return
+        Err(activate_script_command.exec().into())
+    }
+
+    /// Activate the environment interactively by spawning a new shell
+    /// and running the respective activation scripts.
+    ///
+    /// This function should never return as it replaces the current process
+    fn activate_interactive(mut activate_script_command: Command) -> Result<()> {
+        debug!("running activation command: {:?}", activate_script_command);
+
+        // exec should never return
+        Err(activate_script_command.exec().into())
+    }
+
+    /// Used for `eval "$(flox activate)"`
+    fn activate_in_place(mut activate_script_command: Command, shell: ShellWithPath) -> Result<()> {
+        debug!("running activation command: {:?}", activate_script_command);
+
+        let output = activate_script_command
+            .output()
+            .context("failed to run activation script")?;
+        eprint!("{}", String::from_utf8_lossy(&output.stderr));
```

Comment:
Nit: by collecting all of the output and writing them separately, you may get garbled results. Consider the case where `stdout` and `stderr` are interleaved. `stderr` may contain information about what was recently written to `stdout`, but by printing it later you lose that context.

Reply:
Agreed, leaving cause it was a refactor